### PR TITLE
DDO-2879 Support For Azure Service Principal Client Secrets

### PR DESCRIPTION
**Pull request by mflinn-broad**
This PR seeks to achieve two goals. 
1. Add Support for automatic management of Azure Client Secrets in Yale (Azure equivalent of Service Account Keys in Google)
2. Refactor Yale to be more extensible to support rotation of additional types of secrets in the future.

These 2 goals were sufficiently coupled such that it made sense to do both in one big PR.
These code changes touch nearly every single part of Yale's code base however there are no functional changes to existing functionality around GcpSaKeys. When deployed to our existing clusters which only contain GCP SA keys this code change is a no-op. But as the new `AzureClientSecret` CRDs are added to a cluster yale should now be able to process those too.

The vast majority of the logic in yale that is used to process GCP SA Keys is reused to process Azure Client Secrets as Well. 

This was achieved through the introduction of a few Generic Types and interfaces to Yale's code base.
The key ones to be aware of are:

- [YaleCRD](https://github.com/broadinstitute/yale/blob/189a4c80f040319959c19e55ad9011b55d3e2e68/internal/yale/crd/api/v1beta1/types.go#L13) A generic union type that can represent any one of the CRD structs supported by Yale.
- [Identifier](https://github.com/broadinstitute/yale/blob/189a4c80f040319959c19e55ad9011b55d3e2e68/internal/yale/cache/entry.go#L16) An interface that members of yale cache must implement in order to identify themselves to other parts of yale's logic. In the case of GCP Sa Keys Identifier provides access to the service account email and project name, in the case of Azure Client Secrets it provides the Application ID and Tenant ID.
- [Syncable](https://github.com/broadinstitute/yale/blob/189a4c80f040319959c19e55ad9011b55d3e2e68/internal/yale/keysync/keysync.go#L38) An interface type that Yale can use to extract the the relevant data it needs to perform a keysync. Each yale supported CRD type must implement this interface.

## What Parts of the code are non-generic
While now most of Yale's code base is generic over the type of resource it's managing there are a few areas that need to care about the concrete types.
1. The keyops interface must have a backing implementation for each resource type and is now stored as a `map[string]Keyops` in the main Yale struct
2. CRD definitions in `internal/yale/crd` much contain concrete struct types for each supported Yale CRD
3. The points where yale interacts with internal systems. For example marshaling/ unmarshaling k8s crds and secrets, and syncing to k8s secret and vault.
4. Bundle now includes lists of each type of supported crd. There is validation logic that ensures that for a given bundle, only one crd is populated.

## Testing
Unit Tests have been updated to include cases for Azure Client Secrets for all the scenarios currently covered by the GCP SA Key unit tests.

Manual Local testing on a KIND cluster, everything is working well there

Manual testing on static envs. Still in progress.

## Important places to look for reviewers
While this diff is large, a good chunk of it is adding test cases and propagating the new generic and interface types to the existing code. Places warranting eyes for reviewers would be

- [Main yale.go routine](https://github.com/broadinstitute/yale/blob/DDO-2879-azure-support/internal/yale/yale.go)
- [Keysync logic](https://github.com/broadinstitute/yale/blob/DDO-2879-azure-support/internal/yale/keysync/keysync.go)
- [Azure Implementation of Keyops](https://github.com/broadinstitute/yale/blob/DDO-2879-azure-support/internal/yale/keyops/azurekeyops/azurekeyops.go)

### CODE REVIEW

File: internal/yale/keysync/keysync.go [hunk 1]
```diff
@@ -27,7 +29,39 @@ type KeySync interface {
...
 	// Note that this function will update the cache entry's SyncStatus map to reflect any sync's it performs,
 	// but it WILL NOT save the entry to the cache -- that's the caller's responsibility!
-	SyncIfNeeded(entry *cache.Entry, gsks ...apiv1b1.GcpSaKey) error
+	SyncIfNeeded(entry *cache.Entry, gsks []Syncable) error
+}
+
+// Syncable is an interface for objects that can be synced to a Kubernetes secret
+// It is a collection of getter methods that return the information needed to sync to a k8s
+// secret or vault path.
+type Syncable interface {
```
Comment (choover-broad) [inline internal/yale/keysync/keysync.go:38]: Wondering if struct composition could help a bit here? 🤔 

Like if there was a YaleBaseCRD composed into the GCP and Azure crd struct types, and *that* was passed around in packages like this where the details of gcp/azure spec don't matter. That way, no need to implement methods on the GCP and Azure types that are just readers for the struct fields...

File: internal/yale/keyops/azurekeyops/azurekeyops_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
+package azurekeyops
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/broadinstitute/yale/internal/yale/keyops"
+	"github.com/broadinstitute/yale/internal/yale/keyops/azurekeyops/msgraphmock"
```
Comment (choover-broad) [inline internal/yale/keyops/azurekeyops/azurekeyops_test.go:9]: I really appreciate the mocking here!

File: internal/yale/keyops/azurekeyops/azurekeyops_test.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+				SecretText:  &testSecret,
+			})
+	})
+
+	_, _, err := keyOps.Create(testTenantID, testApplicationID)
+	require.Error(t, err)
+	assert.ErrorContains(t, err, "keyId field was nil")
+}
+
+func Test_CreateErrorsIfResponseLacksSecret(t *testing.T) {
```
Comment (choover-broad) [inline internal/yale/keyops/azurekeyops/azurekeyops_test.go:58]: I appreciate how compact and readable these testcases are!

**Review by choover-broad — APPROVED** [review]
this looks great! Really glad to see support for different types of credentials come to Yale :heart: I really appreciate all the work that went into this 💪

File: internal/yale/yale.go [hunk 4]
```diff
@@ -137,80 +167,100 @@ func (m *Yale) processServiceAccount(entry *cache.Entry, gsks []apiv1b1.GcpSaKey
...
+	default:
+		return fmt.Errorf("unknown yaleCRD type %T", yaleCRDs)
+	}
 }
 
-// rotateKey rotates the current active key for the service account, if needed.
-func (m *Yale) rotateKey(entry *cache.Entry, cutoffs cutoff.Cutoffs, gsks []apiv1b1.GcpSaKey) error {
-	rotated, err := m.issueNewKeyIfNeeded(entry, cutoffs, gsks)
+// rotateYaleResource will rotate the active secret for a cache entry if it is time to do so
+func rotateYaleResource[Y apiv1b1.YaleCRD](
```
[the remark this answers is not in the document]
Comment (mflinn-broad) [reply internal/yale/yale.go:195]: The reason I had to do it this way is that go's generics implementation doesn't support generic types on methods :(

File: internal/yale/keysync/keysync.go (hunk 1 above)
Comment (mflinn-broad) [reply internal/yale/keysync/keysync.go:38]: Interesting idea, will ponder it a bit

Comment (mflinn-broad) [reply internal/yale/keysync/keysync.go:38]: I think the catch is that there aren't many places if any where the details of the gcp/azure spec don't matter, but I'll give it a shot

**Pull request merged**